.github/workflows/test.yml

```diff
@@ -1,107 +1,101 @@
-      - uses: actions-rs/cargo@v1
-        with:
-          command: test
-          args: --features=tokio02
-      - uses: actions-rs/cargo@v1
-        with:
-          command: test
-          args: --features=tokio03
 
-  check:
-    name: Check
-    runs-on: ubuntu-latest
-    strategy:
-      matrix:
-        rust:
-          - stable
-          - beta
-          - 1.45.2
     steps:
-      - uses: actions/checkout@v1
-      - uses: actions-rs/toolchain@v1
-        with:
-          profile: minimal
-          toolchain: ${{ matrix.rust }}
-          override: true
-      - uses: actions-rs/cargo@v1
-        with:
-          command: check
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - name: Install rust
+        run: |
+          rustup update --no-self-update stable
+          rustup component add rustfmt
+
+      - name: cargo fmt
+        run: |
+          cargo fmt --all -- --check
 
-  fmt:
-    name: Rustfmt
+  clippy:
+    name: clippy / stable
     runs-on: ubuntu-latest
+
     steps:
-      - uses: actions/checkout@v1
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - name: Install rust
+        run: |
+          rustup update --no-self-update stable
+          rustup component add clippy
+
       - uses: actions-rs/cargo@v1
         with:
-          command: fmt
-          args: --all -- --check
+          command: clippy
+          args: -- -D warnings
 
-  clippy:
-    name: Clippy
+  test:
+    name: test / ${{ matrix.name }}
     runs-on: ubuntu-latest
+
     strategy:
       matrix:
-        rust:
-          - stable
+        include:
+          - name: stable
+            rust: stable
+          - name: beta
+            rust: beta
+          - name: 1.45.2
+            rust: 1.45.2
+
     steps:
-      - uses: actions/checkout@v1
-      - uses: actions-rs/toolchain@v1
-        with:
-          toolchain: ${{ matrix.rust }}
-          override: true
-      - uses: actions-rs/cargo@v1
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - name: Setup cache
+        uses: actions/cache@v2
         with:
-          command: clippy
-          args: -- -D warnings
+          path: |
+            ~/.cargo/registry
+            ~/.cargo/git
+            target
+          key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
+
+      - name: Install rust
+        run: |
+          rustup default ${{ matrix.rust }}
+          rustup update --no-self-update ${{ matrix.rust }}
+
+      - name: Install postfix  
+        run: |
+          DEBIAN_FRONTEND=noninteractive sudo apt-get update
+          DEBIAN_FRONTEND=noninteractive sudo apt-get -y install postfix
 
+      - name: Run SMTP server
+        run: smtp-sink 2525 1000&
+      
+      - name: Install cargo hack
+        run: cargo install cargo-hack --debug
+
+      - name: Test with cargo hack
+        uses: actions-rs/cargo@v1
+        with:
+          command: hack
+          args: test --each-feature
```

Comment:
This could also use run instead of `actions-rs/cargo`

Another thing I realize now, which I hadn't put in the issue, is that --each-feature only tests 1 feature at a time, while we might want to test combinations of multiple features. Something like `cargo hack test --feature-powerset --depth 3` if it doesn't take too long to run

```suggestion
      - name: Test with cargo hack
        run: cargo hack test --each-feature
```